### PR TITLE
Add new backend capability for Riak r_object use

### PR DESCRIPTION
**Pull request by slfritchie**
Add the optional riak_kv backend capability to pass Riak r_object records to/from the backend rather than binary blobs that are serialized/unserialized by `riak_kv_vnode`.

Also, rearrange the QuickCheck tests to permit easier testing of the fs2 backend.

### CODE REVIEW

**Comment by slfritchie** [comment]
NOTE: Branches are apparently hard.  This should be merged to _master_, not the base branch that GitHub is thinking of.

File: rebar.config [hunk 1]
```diff
@@ -1,6 +1,8 @@
 {cover_enabled, true}.
 {edoc_opts, [{preprocess, true}]}.
-{erl_opts, [warnings_as_errors, {parse_transform, lager_transform}]}.
+{erl_opts, [warnings_as_errors,
+            {parse_transform, lager_transform},
+            {d, 'TEST_FS2_BACKEND_IN_RIAK_KV'}]}.
```
Comment (jrwest) [inline rebar.config:5]: what this define is used for?

File: src/riak_kv_vnode.erl [hunk 2]
```diff
@@ -1526,14 +1541,25 @@ object_info({Bucket, _Key}=BKey) ->
     Hash = riak_core_util:chash_key(BKey),
     {Bucket, Hash}.
 
-object_from_binary({B,K}, ValBin) ->
-    object_from_binary(B, K, ValBin).
-object_from_binary(B, K, ValBin) ->
-    case riak_object:from_binary(B, K, ValBin) of
-        {error, R} -> throw(R);
-        Obj -> Obj
+-spec encode_and_Mod_put(
```
Comment (jrwest) [inline src/riak_kv_vnode.erl]: imo this should be named `encode_and_put` to be consistent w/ `do_get_object` which also takes a `Mod`. Also, personally not a fan of capital letters in fn names.

File: src/riak_kv_vnode.erl [hunk 3]
```diff
@@ -1526,14 +1541,25 @@ object_info({Bucket, _Key}=BKey) ->
...
+    case mod_capability_uses_r_object(Mod, ModState, Bucket) of
+        true ->
+            Mod:put_object(Bucket, Key, IndexSpecs, Obj, ModState);
+        false ->
+            ObjFmt = riak_core_capability:get({riak_kv, object_format}, v0),
+            EncodedVal = riak_object:to_binary(ObjFmt, Obj),
+            {Mod:put(Bucket, Key, IndexSpecs, EncodedVal, ModState), EncodedVal}
     end.
 
+mod_capability_uses_r_object(Mod, ModState, Bucket) ->
```
Comment (jrwest) [inline src/riak_kv_vnode.erl]: ELONGFUNNAME ;). maybe shorten a bit: `uses_r_obj(ect)`?

**Comment by jrwest** [comment]
`backend_eqc` still passes (using mem. backend) with changes. ran a few riak tests (like `verify_riak_object_reformat` and `mapred_rt_verify`) against it and all passed. Changes seem fine. Would like to see couple of style comments I left addressed but other than that +1. 

EDIT: any reason to not re-submit this PR against master?

File: rebar.config (hunk 1 above)
Comment (slfritchie) [inline rebar.config:5]: For testing riak_kv_fs2_backend.erl, which is in the CS repo.

**Comment by slfritchie** [comment]
Closing with commit 146e3ca047139311ff705e74def7dee2ab97a6c4 (the result of a merge conflict, which I fixed, EUnit tests passing)

**Pull request closed without merging**